llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -10621,12 +10621,64 @@ bool SIInstrInfo::analyzeCompare(const MachineInstr &MI, Register &SrcReg,
   return false;
 }
 
+bool SIInstrInfo::invertSCCUse(MachineInstr *SCCDef) const {
+  MachineBasicBlock *MBB = SCCDef->getParent();
+  const SIRegisterInfo *TRI = ST.getRegisterInfo();
+  SmallVector<MachineInstr *, 2> InvertInstr;
+  bool SCCIsDead = false;
+
+  // Scan instructions for SCC uses that need to be inverted until SCC is dead.
+  for (MachineInstr &MI :
+       make_range(std::next(MachineBasicBlock::iterator(SCCDef)), MBB->end())) {
+    if (MI.findRegisterUseOperandIdx(AMDGPU::SCC, TRI, false) != -1) {
+      if (MI.getOpcode() == AMDGPU::S_CSELECT_B32 ||
+          MI.getOpcode() == AMDGPU::S_CSELECT_B64 ||
+          MI.getOpcode() == AMDGPU::S_CBRANCH_SCC0 ||
+          MI.getOpcode() == AMDGPU::S_CBRANCH_SCC1)
+        InvertInstr.push_back(&MI);
+      else
+        return false;
+    }
+    if (MI.modifiesRegister(AMDGPU::SCC, TRI)) {
+      SCCIsDead = true;
+      break;
+    }
+  }
+
+  const MachineRegisterInfo &MRI =
+      SCCDef->getParent()->getParent()->getRegInfo();
+  // If SCC is still live, verify that it is not live past the end of this
+  // block.
+  if (!SCCIsDead && MRI.tracksLiveness())
+    SCCIsDead = MBB->computeRegisterLiveness(TRI, AMDGPU::SCC, MBB->end(), 0) ==
```

Comment:
I'm not sure what a threshold of 0 is supposed to mean? Is this equivalent to just checking the dead flag on the def you already have?

Reply:
Dead is not enough.  Kill is not available.  Added new `MBB->isLiveOut`.

Reply:
What do you mean by "kill is not available"? There is no _guarantee_ that kill flags will be set, it is just a best effort thing. Maybe we could improve some other pass to set the kill flag in more cases?

Reply:
> What do you mean by "kill is not available"? There is no _guarantee_ that kill flags will be set, it is just a best effort thing. Maybe we could improve some other pass to set the kill flag in more cases?

Sorry I wasn't clear.  I meant that the best effort in this case is not providing much information.  In the lit tests and building the check-libc-amdgcn-amd-amdhsa target, SCC was never determined to be dead because of the kill flag.  SCC was determined to be dead with this distribution:

In lit testing:
  !liveOut SCC 16

In check-libc-amdgcn-amd-amdhsa:
  !liveOut SCC 281
  MI.definesRegister(SCC) 168